app/lib/logic.js

```
@@ -68,31 +97,117 @@ async function updateInsurancePolicy(tx) { // eslint-disable-line no-unused-vars
+    const factory = getFactory();
+    const namespace = 'org.example.mynetwork';
 
-    const assetRegistry = await getAssetRegistry(tx.record.getFullyQualifiedType());
-    await assetRegistry.update(tx.record)
+    const claim = factory.newResource(namespace, 'InsuranceClaim', tx.claimId);
+    claim.claimerType = tx.claimerType;
+    claim.provider = tx.provider;
+    claim.patient = tx.patient;
+
+    // TODO: more logistics to check the validity of the insurance policy (e.g. is it expired?)
+
+    if (tx.policy.insuree.getIdentifier() != tx.patient.getIdentifier()) {
+        throw new Error('Insurance policy insuree not same as insurance claim patient');
+    }
+    claim.policy = tx.policy;
+    
+    if (tx.record !== undefined) {
+        if (tx.record.provider.getIdentifier() != tx.provider.getIdentifier()) {
+            throw new Error('Medical record provider not same as insurance claim provider');
+        }
+        if (tx.record.patient.getIdentifier() != tx.patient.getIdentifier()) {
+            throw new Error('Medical record patient not same as insurance claim patient');
+        }
+        claim.record = tx.record;
+    }
+
+    const assetRegistry = await getAssetRegistry(claim.getFullyQualifiedType());
+    await assetRegistry.add(claim);
 }
 
 /**
- * Unauthorize a Particitant to revoke read access to a Medical Record
- * @param {org.example.mynetwork.UnauthorizeMedicalRecord} unauthorizeMedicalRecord
+ * Update Insurance Claim
+ * @param {org.example.mynetwork.UpdateInsuranceClaim} updateInsuranceClaim - the UpdateInsuranceClaim transaction
  * @transaction
  */
-async function unauthorizeMedicalRecord(tx) {
-    const authorizedParties = tx.record.authorizedParties.filter(
-        authorizedParty => authorizedParty.getIdentifier() !== tx.unauthorizedParty.getIdentifier() 
-    );
+async function updateInsuranceClaim(tx) {
+    const claim = tx.claim;
+
+    if (claim.claimStatus != 'SUBMITTED') {
```

I think it's better to define an enum claimStatus containing 'SUBMITTED', 'REJECTED' and so on somewhere instead of hardcoding it in the transaction

---

Yes it is a `enum` type for `claimStaus` field. But it is still a `string` for each `type` in the `enum` and the interpreter will know whether it is a valid type.

---

https://github.com/ecefydp2019g54/BeyondP2P/blob/76d81cf9410070e5d1cecf4878455a2616aea08f/app/models/org.example.mynetwork.cto#L37-L41

---

claim.claimStatus != STATUS.SUBMITTED

Something like this?

---

Let's deal with this thing in the future, I will merge it for now and start renaming namespace